pkg/calculator/experimentcalc/experiment_calculator.go

```diff
@@ -0,0 +1,650 @@
+		Cursor:               "",
+		EnvironmentNamespace: namespace,
+		Statuses: []experiment.Experiment_Status{
+			experiment.Experiment_WAITING,
+			experiment.Experiment_RUNNING,
+		},
+	}
+	resp, err := e.experimentClient.ListExperiments(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Experiments, nil
+}
+
+func (e ExperimentCalculator) getEvaluationCount(
+	ctx context.Context,
+	req *eventcounter.GetExperimentEvaluationCountRequest,
+) (map[string]*eventcounter.VariationCount, error) {
+	variationCounts := make(map[string]*eventcounter.VariationCount)
+	evaluationCount, err := e.eventCounterClient.GetExperimentEvaluationCount(ctx, req)
+	if err != nil {
+		return variationCounts, err
+	}
+	for _, variationCount := range evaluationCount.VariationCounts {
+		variationCounts[variationCount.VariationId] = variationCount
+	}
+	return variationCounts, nil
+}
+
+func (e ExperimentCalculator) getGoalCount(
+	ctx context.Context,
+	req *eventcounter.GetExperimentGoalCountRequest,
+) (map[string]*eventcounter.VariationCount, error) {
+	variationCounts := make(map[string]*eventcounter.VariationCount)
+	goalCount, err := e.eventCounterClient.GetExperimentGoalCount(ctx, req)
+	if err != nil {
+		return variationCounts, err
+	}
+	for _, variationCount := range goalCount.VariationCounts {
+		variationCounts[variationCount.VariationId] = variationCount
+	}
+	return variationCounts, nil
+}
+
+func (e ExperimentCalculator) calcGoalResult(
+	ctx context.Context,
+	evalVariationCounts,
+	goalVariationCounts map[string]*eventcounter.VariationCount,
+	baseVid string,
+) *eventcounter.GoalResult {
+	goalResult := &eventcounter.GoalResult{}
+	length := len(goalVariationCounts)
+	vids := make([]string, 0, length)
+	goalUc, evalUc := make([]int64, 0, length), make([]int64, 0, length)
+	vrs := make(map[string]*eventcounter.VariationResult, length)
+	valueMeans, valueVars := make([]float64, 0, length), make([]float64, 0, length)
+	baselineIdx, loopIdx := 0, 0
+
+	for vid, goalVariationCount := range goalVariationCounts {
+		if _, ok := evalVariationCounts[vid]; !ok {
+			return goalResult
+		}
+		vids = append(vids, vid)
+		evalVariationCount := evalVariationCounts[vid]
+		vr := &eventcounter.VariationResult{
+			VariationId:     vid,
+			ExperimentCount: copyVariationCount(goalVariationCount),
+			EvaluationCount: copyVariationCount(evalVariationCount),
+		}
+		evalUc = append(evalUc, evalVariationCount.UserCount)
+		goalUc = append(goalUc, goalVariationCount.UserCount)
+		valueMeans = append(valueMeans, goalVariationCount.ValueSumPerUserMean)
+		valueVars = append(valueVars, goalVariationCount.ValueSumPerUserVariance)
+
+		goalResult.VariationResults = append(goalResult.VariationResults, vr)
+		vrs[vid] = vr
+		if vid == baseVid {
+			baselineIdx = loopIdx
+		}
+		loopIdx++
+	}
+	// Skip the calculation if evaluation count is less than goal count.
+	for i := 0; i < len(evalUc); i++ {
+		if evalUc[i] < goalUc[i] {
+			return goalResult
```

Comment:
Let's add a new metric to know that the calculation was skipped.

```go
skippedEvaluationCountLessThanGoalEvent
```

Also, if we skip the calculation, should it return `&eventcounter.GoalResult{}` instead?
Because at [this line](https://github.com/bucketeer-io/bucketeer/pull/395/files#diff-6b2e151d6e7453bb8a10c3b450be4ba51fc681fe35b1b485d43d452d0deb19b5R300) it is happening results.

Reply:
> Also, if we skip the calculation, should it return `&eventcounter.GoalResult{}` instead? Because at [this line](https://github.com/bucketeer-io/bucketeer/pull/395/files#diff-6b2e151d6e7453bb8a10c3b450be4ba51fc681fe35b1b485d43d452d0deb19b5R300) it is happening results.

The original Python code logic was:
 [line 1](https://github.com/bucketeer-io/bucketeer/blob/main/python/src/lib/calculator/job/calculate_experiments.py#L459)
[line 2](https://github.com/bucketeer-io/bucketeer/blob/main/python/src/lib/calculator/job/calculate_experiments.py#L434)

I think we can return the empty `GoalResult{}` here, but also you can confirm this logic with the original writer.

Anyway, we should add this metric first.

Reply:
Next week the person who wrote the Python code will take some time to review this part.
So we can confirm with him the best approach for those cases. For now, let's just add the metrics.

Reply:
IMHO, the return values should be `goalResult` not empty struct because `goalResult` must contain the count values until here. The cause of this error can be a network error, client error, system error, or others, ending up the evaluation count is less than the goal count. This does not supposed to happen usually, but it can happen. I think we should return the real values to end-users rather than returning nothing. wdyt?

Reply:
@kentakozuka, I am concerned about returning the unfinished result, affecting the visualization on the UI, and introducing bugs that are hard to debug later.

Since the calculator will calculate the probability again in the next job, I thought it would be better to treat it as an error and not update the experiment result. Also, as you said, this usually won't happen, so implementing it not to update the experiment result sounds safer.

Let me know if I'm missing something.

Reply:
I think that we should define what we treat as error and what not.
It can come to this line even if the Bucketeer works as we expected (and not sure it would not come again in the next job. In this line we can still return;
- evalUc
- goalUc
- valueMeans
- valueVars

Also, would it be better to return either?;
- `goalResult`, nil
- `nil, err`

Reply:
@kentakozuka, the appended data [here](https://github.com/bucketeer-io/bucketeer/blob/e6180757e36ee6bb6558ed5d9ebea433b7dabed8/pkg/calculator/experimentcalc/experiment_calculator.go#L309) is the data display here?

If so, it should be ok to return the `goalResult`. 
If the goal count is higher than the evaluation count, the UI will only display the graph below, right?

![Screenshot 2023-06-28 at 14 47 54](https://github.com/bucketeer-io/bucketeer/assets/2486691/8da56cef-fe99-459d-920f-0c57b3da1d45)

Reply:
> he appended data here is the data displayed in the screenshot below?

I'm not sure but possibly yes.
https://github.com/bucketeer-io/bucketeer/blob/b3a0abc1385f9ec79d954aa4b7d8a7d11085a9c7/ui/web-v2/apps/admin/src/components/GoalResultDetail/index.tsx#L137